packages/nhost-js/src/core/nhost-client.ts

```diff
@@ -4,10 +4,31 @@ import { HasuraStorageClient } from '@nhost/hasura-storage-js'
 import { NhostFunctionsClient } from '../clients/functions'
 import { NhostGraphqlClient } from '../clients/graphql'
 
-export type NhostClientConstructorParams = Omit<NhostAuthConstructorParams, 'url'> & {
+type Params = Omit<NhostAuthConstructorParams, 'url'> & {
+  /** Nhost backend url
+   * @example https://my-app.nhost.run
+   */
   backendUrl: string
+  /** GraphQL endpoint. If given, it will take precedence over `backendUrl` */
+  graphqlUrl?: string
+  /** Hasura Auth endpoint. If given, it will take precedence over `backendUrl` */
+  authUrl?: string
+  /** Hasura Storage endpoint. If given, it will take precedence over `backendUrl` */
+  storageUrl?: string
+  /** Functions endpoint. If given, it will take precedence over `backendUrl` */
+  functionsUrl?: string
 }
 
+type RequireAtLeastOne<T, Keys extends keyof T = keyof T> = Pick<T, Exclude<keyof T, Keys>> &
+  {
+    [K in Keys]-?: Required<Pick<T, K>> & Partial<Pick<T, Exclude<Keys, K>>>
+  }[Keys]
+
+export type NhostClientConstructorParams = RequireAtLeastOne<Params, 'backendUrl' | 'graphqlUrl'> &
```

Comment:
Wouldn't it make more sense to make sure that `backendUrl` is required, so that we could have it as a fallback value for any of the missing values? Based on `type Params` it's seemingly required, but `NhostClientConstructorParams` makes it optional.

Reply:
This way we could make our constructor less error prone by simply just using the fallback value if any of the custom URLs are missing. What do you think about it?

Reply:
the logic is that either `backendUrl` is present, or all of the `graphqlUrl` etc are present.
The typing is not ideal but it works.

In the case of custom endpoints, there is no server to fall back to.
If we enforce `backendUrl` while wants to define their own custom endpoints, they would still give a value that would not be used eventually to pass the typings error, leading to potential confusion.